cli/src/commands/scan.rs

```diff
@@ -641,6 +629,57 @@ impl ScanState {
     }
 }
 
+// Process scan results and output matches, non-matches, or count of matches
+// based upon command line arguments. Return the number of "matched" files so
+// the state can be updated.
+fn process_scan_results(
+    args: &ArgMatches,
+    file_path: &Path,
+    scan_results: &ScanResults,
+    output: &Sender<Message>,
+) -> usize {
+    let negate = args.get_flag("negate");
+    let count = args.get_flag("count");
+
+    if count {
+        // The behavior of original YARA is to ignore things like -n and
+        // -t when using -c so we are doing it here also.
+        let match_count = scan_results.matching_rules().len();
+        print_match_count(args, file_path, &match_count, output);
+        return match_count;
+    } else if negate {
+        let mut matching_rules = scan_results.non_matching_rules();
+        let match_count = matching_rules.len();
+        print_matching_rules(args, &file_path, &mut matching_rules, output);
+        return match_count;
+    } else {
+        let mut matching_rules = scan_results.matching_rules();
+        let match_count = matching_rules.len();
+        print_matching_rules(args, &file_path, &mut matching_rules, output);
+        return match_count;
+    };
```

Comment:
In this implementation `--negate` can't be use used in conjunction with `--count`, you either use one or the other. However, I don't see a reason for that, I think that they should be allowed to be used together. Also, when you use `--negate`, the number returned by `--count` should be the number of non-matching rules. In other words, `--negate` should invert the results, when used without `--count` it returns the non-matching rules, and when used with `--count` it should return the number of non-matching rules.

Reply:
Thanks! I was trying to mimic the behavior of original YARA here, but I like you're suggestion better.